### DISCUSSION TITLE
ArgumentOutOfRangeException for path at runtime

### DISCUSSION THREAD

**Discussion by Mar-he** [serenity-is/Serenity#6682] [Help] [answered]
I have inherited a Serenity application, based on .net 3 and Serenity 3.13.
I've managed to get it building, after finding reading up on `sergen` and its restore command. Now I am faced with a weird runtime exception, that does not make sense to me - yet. I'm hoping you can shed some light on it.

```cs
ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'path')
System.Web.Hosting.HostingEnvironment.MapPath(string path)
Serenity.Web.ContentHashCache.ResolveWithHash(string contentUrl)
Serenity.Web.HtmlScriptExtensions.ScriptBundle(IHtmlHelper helper, string bundleKey)
AspNetCoreGeneratedDocument.Views_Shared__LayoutHead.ExecuteAsync() in _LayoutHead.cshtml
+
@Html.ScriptBundle("Base")
Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, bool invokeViewStarts)
Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
Microsoft.AspNetCore.Mvc.TagHelpers.PartialTagHelper.RenderPartialViewAsync(TextWriter writer, object model, IView view)
Microsoft.AspNetCore.Mvc.TagHelpers.PartialTagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)
Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.<RunAsync>g__Awaited|0_0(Task task, TagHelperExecutionContext executionContext, int i, int count)
AspNetCoreGeneratedDocument.Views_Shared__LayoutNoNavigation.<ExecuteAsync>b__9_0() in _LayoutNoNavigation.cshtml
+
    <partial name="@MVC.Views.Shared._LayoutHead" />
Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperExecutionContext.SetOutputContentAsync()
AspNetCoreGeneratedDocument.Views_Shared__LayoutNoNavigation.ExecuteAsync() in _LayoutNoNavigation.cshtml
+
<html lang="@System.Globalization.CultureInfo.CurrentUICulture.Name"@Html.Raw(rtl ? " dir=\"rtl\"" : "")>
```

Trying to dig into the problem, I've set a breakpoint at `@Html.ScriptBundle("Base")` in my _LayoutHead.cshtml. Single stepping through the call stack it breaks while trying to resolve `IHttpContextAccessor` inside `DependencyResolver.Resolve`, although it resolves it 3 times successfully before, so I guess this is just a red herring.

At this point I am unable to determine which path it is failing on.

**Comment by Mar-he** [comment, marked as answer]
**tl;dr**: in `wwwroot/Scripts/site/ScriptBundels.json` there was a j`query-3.3.1..js` entry - which causes .net core to throw an exception.

I found the solution, by switching from Visual Studio to Rider and dive into the sources of .net and Serenity.
The ScriptBundle("Base") call, was invoking `Serenity.Web.HtmlScriptExtensions.ScriptBundle`:

```cs
   public static HtmlString ScriptBundle(this IHtmlHelper helper, string bundleKey)
    {
      if (helper == null)
        throw new ArgumentNullException(nameof (helper));
      if (string.IsNullOrEmpty(bundleKey))
        throw new ArgumentNullException(nameof (bundleKey));
      HttpContext httpContext = helper.ViewContext.HttpContext;
      if (ScriptBundleManager.IsEnabled)
        return helper.Script("dynamic://Bundle." + bundleKey);
      StringBuilder stringBuilder = new StringBuilder();
      foreach (string bundleInclude in ScriptBundleManager.GetBundleIncludes(bundleKey))
      {
        if (!string.IsNullOrEmpty(bundleInclude))
        {
          string contentUrl = bundleInclude == null || !bundleInclude.StartsWith("dynamic://", StringComparison.OrdinalIgnoreCase) ? VirtualPathUtility.ToAbsolute(BundleUtils.ExpandVersionVariable(bundleInclude)) : VirtualPathUtility.ToAbsolute("~/DynJS.axd/" + DynamicScriptManager.GetScriptInclude(bundleInclude.Substring(10)));
          HashSet<string> includedScripts = HtmlScriptExtensions.GetIncludedScripts(httpContext);
          if (!includedScripts.Contains(contentUrl))
          {
            includedScripts.Add(contentUrl);
            stringBuilder.AppendLine(string.Format("    <script src=\"{0}\" type=\"text/javascript\"></script>\n", (object) WebUtility.HtmlEncode(ContentHashCache.ResolveWithHash(contentUrl))));
          }
        }
      }
      return new HtmlString(stringBuilder.ToString());
    }
```

This was calling `System.Web.Hosting.HostingEnvironment.MapPath(string path)` on line 100 (`stringBuilder.AppendLine(string.Format("    <script src=\"{0}\" type=\"text/javascript\"></script>\n", (object) WebUtility.HtmlEncode(ContentHashCache.ResolveWithHash(contentUrl))));`)

`MapPath` has the following guard clause on line 20: `if (path.IndexOf("..") < 0 && path.IndexOf("//") < 0 && path.IndexOf(":/") < 0)`, which was throwing the exception. 

Now that I knew which filepath was the issue, I only need to find the place, were this was configured.
